File: js/main3.js
// Problem: TCP Handshake with Incorrect State Transitions
// Points: 8/10
// Expected Output: SYN sent, Connection Established, ESTABLISHED
class TCP {
    constructor() {
        this.state = "CLOSED";
    }

    send_SYN() {
        if (this.state === "CLOSED") {
            this.state = "SYN_SENT";
            console.log("SYN sent");
        }
    }

    recv_SYN_ACK() {
        if (this.state === "SYN_SENT") {
            this.state = "ESTABLISHED"; 
            console.log("Connection Established");
        }
    }

    send_ACK() {
        if (this.state === "ESTABLISHED") {
            console.log("ACK Sent, connection stable.");
        }
    }
}

let connection = new TCP();
connection.send_SYN();
connection.recv_SYN_ACK();
console.log(connection.state);
connection.send_ACK();

File: py/main3.py
# Problem: TCP Congestion Control with Broken Window Logic
# Points: 7/10
# Expected Output: 1
class TCP:
    def __init__(self):
        self.cwnd = 1

    def packet_acknowledged(self):
        self.cwnd *= 2  

    def packet_lost(self):
        self.cwnd /= 2  

tcp = TCP()
tcp.packet_acknowledged()
tcp.packet_acknowledged()
tcp.packet_lost()
print(tcp.cwnd)
